services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/update/UpdateExpression.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.enhanced.dynamodb.update;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.utils.CollectionUtils;
+
+/**
+ * Contains sets of {@link UpdateAction} that represent the four DynamoDB update actions: SET, ADD, REMOVE and DELETE.
+ * <p>
+ * Use this class to build an immutable UpdateExpression with one or more UpdateAction. An UpdateExpression may be merged
+ * with another. When two UpdateExpression are merged, the actions of each group of UpdateAction, should they exist, are
+ * combined; all SET actions from each expression are concatenated, all REMOVE actions etc.
+ * <p>
+ * DynamoDb Enhanced will convert the UpdateExpression to a format readable by DynamoDb,
+ * <p>
+ * Examples of adding actions for attributes directly:
+ * <pre>
+ * {@code
+ * RemoveUpdateAction removeAction = ...
+ * List<SetUpdateAction> setActions = ...
+ * UpdateExpression.builder()
+ *                 .addAction(removeAction)
+ *                 .actions(setActions)
+ *                 .build();
+ * }
+ * </pre>
+ *
+ * See respective subtype of {@link UpdateAction}, for example {@link SetUpdateAction}, for details on creating that action.
+ */
+@SdkPublicApi
+public final class UpdateExpression {
+
+    private final List<RemoveUpdateAction> removeActions;
+    private final List<SetUpdateAction> setActions;
+    private final List<DeleteUpdateAction> deleteActions;
+    private final List<AddUpdateAction> addActions;
+
+    private UpdateExpression(Builder builder) {
+        this.removeActions = builder.removeActions;
+        this.setActions = builder.setActions;
+        this.deleteActions = builder.deleteActions;
+        this.addActions = builder.addActions;
+    }
+
+    /**
+     * Constructs a new builder for {@link UpdateExpression}.
+     *
+     * @return a new builder.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public List<RemoveUpdateAction> removeActions() {
+        return Collections.unmodifiableList(removeActions);
```

Comment:
I think we should create a new List and wrap it with unmodifiableList in the ctor. Can we do a find and replace to fix all occurrences of list and map?